### PR TITLE
[ONNX] Support custom axis name through dynamic_shapes

### LINKED ISSUE #143443: [ONNX] Rename dynamic shapes produced by ExportedProgram to dynamic_axes
`torch.export.export` names dynamic shapes to be s0, s1, s2, s3, ..., etc. However, in ONNX, users could pass in the naming through `dynamic_axes` and `input_names`. We need to rename them to what users request.

### PR DESCRIPTION
**Pull request by titaiwangms**
Fixes #143443

This PR aims to support custom dynamic axis naming through dynamic_shapes. Currently, _Dim and _DimHint do not support dynamic axis naming (#144273).

1. **the original dynamic shapes guarantee**
The axis renaming is only applied when dynamic shapes include string instead of all _Dim and _DimHint. Thus, there will not be any inconsistent behavior to dynamic_shapes with torch.export.export if the given dynamic shapes follow torch.export.export format.
2. _DimHint.AUTO is applied to the axes that are specified with custom names to avoid exporter crash. (_DimHint.DYNAMIC crashes when the export fails.)
3.  There's no need to handle cases where kwargs are out of order with the model signature,
    as torch.export.export supports dynamism only when kwargs and dynamic_shapes are provided in order.
    https://github.com/pytorch/pytorch/blob/49082f9dba3b79a344cb03652972ddbe7c3729cc/torch/export/_trace.py#L2034
4. If `torch.onnx.ExportedProgram` finds the axes share the same constraints, they will have the same name (e.g. s0, s1, ...). Therefore, even if the ONNX users specify them with different custom names, they won't be respected.

Example model:
```python
        class NestedModel(torch.nn.Module):
            def forward(
                self,
                x: torch.Tensor,
                ys: list[torch.Tensor],
                zs: dict[str, torch.Tensor],
                c: torch.Tensor,
            ):
                y = ys[0] + ys[1] + zs["a"] + zs["b"]
                w = 5
                if x.shape[0] < 3 and c.shape[0] != 4:
                    return x + w, x + y, c
                else:
                    return x - w, x - y, c

        input = (
            torch.ones(5),
            [torch.zeros(5), torch.ones(5)],
            {"a": torch.zeros(5), "b": torch.ones(5)},
            torch.ones(6),
        )

        dynamic_shapes = (
            {0: torch.export.Dim("dim_x", min=3)},  # _Dim
            [("custom_name_axis_ys_0",), (torch.export.Dim.AUTO,)],  # custom name
            {
                "a": {0: torch.export.Dim.AUTO},
                "b": ("custom_name_axis_zs_b_0",),
            },  # _DimHint
            {0: "custom_name_axis_c_0"},  # custom name
        )

```

### CODE REVIEW

File: torch/onnx/_internal/exporter/_compat.py [hunk 1]
```diff
@@ -237,21 +84,27 @@ def export_compat(
...
+                        input_names=input_names,
+                        output_names=set(output_names or ()),
+                    )
                 )
             except Exception as e:
                 raise RuntimeError(
                     "# Failed to convert 'dynamic_axes' to 'dynamic_shapes'. "
                     "Please provide 'dynamic_shapes' directly. "
-                    "Refer to the documentation for 'torch.export.export' for more information on dynamic shapes."
+                    "Refer to the documentation of 'torch.export.export' for more information on dynamic shapes."
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_compat.py:101]: "for" seems good?

**Review by justinchuby — APPROVED** [review]

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                if isinstance(dim, str):
+                    return True
+        elif isinstance(axes, (list, tuple)):
+            for dim in axes:
+                if isinstance(dim, str):
+                    return True
+    return False
+
+
+def convert_str_to_export_dim(
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_dynamic_shapes.py:172]: Possible to have unit tests for this function?

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+
+
+def _get_custom_axis_name(axis: _Dim | str) -> str:
+    """Get the custom axis name from a torch.export.Dim."""
+    if isinstance(axis, _Dim):
+        return axis.__name__
+    return axis
+
+
+def iterate_and_change_axis_names(
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_dynamic_shapes.py:291]: Possible to have unit tests for this function? Also this function is fitting in ir_passes.py

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        return inspect.signature(should_be_callable)
+    raise ValueError("model has no forward method and is not callable")
+
+
+def _all_values(model_or_function: ir.Model | ir.Function):
+    """Yield all values in a model."""
+    if isinstance(model_or_function, ir.Function):
+        yield from model_or_function.inputs
+        for node in ir.traversal.RecursiveGraphIterator(model_or_function):
+            yield from node.outputs
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_dynamic_shapes.py:367]: Simply return all function values from a model?

File: torch/onnx/_internal/exporter/_onnx_program.py [hunk 5]
```diff
@@ -254,6 +255,20 @@ def release(self) -> None:
...
+    ) -> None:
+        """Rename dynamic axes in a model according to the specified dynamic_axes names."""
+        rename_mapping = _dynamic_shapes.create_rename_mapping(
+            self.model.graph.inputs, dynamic_shapes
+        )
+        _dynamic_shapes.iterate_and_change_axis_names(self.model, rename_mapping)
+        for model_function in self.model.functions.values():
+            _dynamic_shapes.iterate_and_change_axis_names(
+                model_function, rename_mapping
+            )
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_onnx_program.py:270]:
```suggestion
        _dynamic_shapes.iterate_and_change_axis_names(self.model, rename_mapping)
```

model the function iteration into the iterate_and_change_axis_names pass

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    model,
+    args: tuple[Any, ...],
+    kwargs: dict[str, Any] | None,
+    *,
+    dynamic_axes=None,
+    output_names: set[str],
+    input_names: Sequence[str] | None = None,
+) -> tuple[dict[str, Any | None] | None, tuple[Any, ...], dict[str, Any] | None]:
+    """
+    Converts dynamic_axes into dynamic_shapes by wrapping the axis names with torch.export.Dim.AUTO.
```
Comment (xadupre) [inline torch/onnx/_internal/exporter/_dynamic_shapes.py:27]: ``torch.export.Dim.AUTO``

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+    dynamic_axes examples:
+    (1) dynamic_axes = {"x": {0: "my_custom_axis_name_1"}, "y": {1: "my_custom_axis_name_2"}}
+    (2) dynamic_axes = {"x": [0], "y": [1]}
+
+    these will be converted to dynamic_shapes respectively:
+    (1) dynamic_shapes = {"x": {0: Dim.AUTO}, "y": {1: Dim.AUTO}}
+    (2) dynamic_shapes = {"x": {0: Dim.AUTO}, "y": {1: Dim.AUTO}}
+
+    Detail on Dim.AUTO: https://github.com/pytorch/pytorch/pull/133620
```
Comment (xadupre) [inline torch/onnx/_internal/exporter/_dynamic_shapes.py:37]:
```
`#133620 <https://github.com/pytorch/pytorch/pull/133620>`_
```

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+    if len(input_names) < len(flat_dynamic_shapes):
+        raise ValueError(
+            "To construct dynamic_axes from dynamic_shapes, "
+            f"number of input names ({len(input_names)}) should be greater than or equal to "
+            f"the number of graph inputs(flat) ({len(flat_dynamic_shapes)})"
+        ) from exception
+
+    dynamic_axes: dict[str, list[int]] = {}
+    # input names are assigned in order
+    for input_name, axes in zip(input_names, flat_dynamic_shapes):
```
Comment (xadupre) [inline torch/onnx/_internal/exporter/_dynamic_shapes.py:135]: What about nested list or dictionary?

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+            UserWarning,
+            stacklevel=3,
+        )
+        return {}
+    rename_mapping: dict[str, str] = {}
+    # NOTE: We assume that the flat_dynamic_shapes is in the same order as the inputs
+    # When the axis is static, or it connects to _DimHint in dynamic shapes, we skip renaming
+    for idx, axes in enumerate(flat_dynamic_shapes):
+        input = inputs[idx]
+        if isinstance(axes, dict):
```
Comment (xadupre) [inline torch/onnx/_internal/exporter/_dynamic_shapes.py:241]: What about nested list or dictionary?

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+# mypy: allow-untyped-defs
+from __future__ import annotations
+
+import inspect
+import re
+import warnings
+from typing import Any, Sequence
+
+import torch
+from torch.export.dynamic_shapes import _Dim, _DimHint
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_dynamic_shapes.py:12]: Is it possible to import modules only?

File: torch/onnx/_internal/exporter/_dynamic_shapes.py (hunk 9 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py:241]: The list and dict here are leaves that are kept from flatten in line 227. Please check the function `_flatten_dynamic_shapes_to_axes`. That is, at this point, there will not be any nested list and dict.

File: torch/onnx/_internal/exporter/_dynamic_shapes.py (hunk 10 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py:12]: I don't want to use dynamic_shapes._Dim, so I didn't import modules. The module name dynamic_shapes is too general. Do you think 

```python
from torch.export import dynamic_shapes as export_dynamic  # or some other recommendations?
```

is better than importing _Dim and _DimHint?
I don't have preference, just didn't like dynamic_shapes

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+def _flatten_dynamic_shapes_to_axes(
+    dynamic_shapes: dict[str, Any | None] | tuple[Any, ...] | list[Any],
+) -> tuple[list[Any], _pytree.TreeSpec]:
+    # If it's a dict/list/tuple with torch.export._Dim, we consider it's an axis to dim mapping
+    def is_axes(x) -> bool:
+        return (
+            isinstance(x, dict)
+            and all(
+                isinstance(k, int)
+                and (v is None or isinstance(v, (_Dim, _DimHint, str)))
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py:336]: It's covered: https://github.com/pytorch/pytorch/blob/206ad9f4ad0d342edbabc7bae3f63e3d3909e31f/torch/export/dynamic_shapes.py#L142

Or there is another instance you are referring to?

File: torch/onnx/_internal/exporter/_dynamic_shapes.py (hunk 8 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py:135]: Sam as the comment below.

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+    dynamic_shapes_with_export_dim = _pytree.tree_unflatten(
+        dynamic_shapes_with_export_dim, tree_structure
+    )
+    return (
+        dynamic_shapes_with_export_dim,
+        True,
+    )
+
+
+def create_rename_mapping(
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py:218]: This involves to create ir.Inputs and it's only about mapping. Unless you have some corner cases in mind that we should keep an eye on? I can make that a test.

File: torch/onnx/_internal/exporter/_dynamic_shapes.py (hunk 3 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py:291]: Do you recommend to delete the method from ONNXProgram? We can have it as a pass, but we will need to accpet the whole _dynamic_shapes related processings happening inside _compat.py, which is already crowded with step by step validation.

File: torch/onnx/_internal/exporter/_dynamic_shapes.py (hunk 4 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py]: Done

File: torch/onnx/_internal/exporter/_onnx_program.py (hunk 5 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_onnx_program.py]: Done

File: torch/onnx/_internal/exporter/_dynamic_shapes.py (hunk 2 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py:172]: Done

File: torch/onnx/_internal/exporter/_dynamic_shapes.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py:291]: I think we can keep the method. I would just mode the model transform part as a pass

File: torch/onnx/_internal/exporter/_dynamic_shapes.py (hunk 10 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py:12]: Feel free to keep the original, thanks!

File: torch/onnx/_internal/exporter/_dynamic_shapes.py (hunk 3 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_dynamic_shapes.py]: Done. Now it calls the pass inside the method. I find making a test for this pass requires a lot effort because I need to build a model with all sort of testing conditions. Especially when we already have e2e tests basically guarding the whole functionality. Do you still think we should have a particular test on this? If so, is there an easy way to do it?

File: torch/onnx/_internal/exporter/_ir_passes.py [hunk 13]
```diff
@@ -30,6 +31,53 @@ def rename_outputs(model: ir.Model, new_names: Sequence[str]) -> None:
...
+
+
+def _replace_names(shape_expr: str, rename_mapping: dict[str, str]) -> str:
+    """Replace all known names in a shape expression with new names."""
+    for old_name, new_name in rename_mapping.items():
+        shape_expr = re.sub(rf"\b{old_name}\b", new_name, shape_expr)
+    return shape_expr
+
+
+def rename_axis(model: ir.Model, rename_mapping: dict[str, str]) -> None:
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_ir_passes.py:55]: I would check a few cases when the mapping contains strings that is not in the model; I would also have a test the ensure that shapes in functions are renamed.

File: torch/onnx/_internal/exporter/_ir_passes.py [hunk 14]
```diff
@@ -30,6 +31,53 @@ def rename_outputs(model: ir.Model, new_names: Sequence[str]) -> None:
...
+    for node in ir.traversal.RecursiveGraphIterator(model.graph):
+        yield from node.outputs
+    # Yield all values in functions
+    for function in model.functions.values():
+        yield from function.inputs
+        for node in ir.traversal.RecursiveGraphIterator(function):
+            yield from node.outputs
+
+
+def _replace_names(shape_expr: str, rename_mapping: dict[str, str]) -> str:
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_ir_passes.py:48]: Done.

I had to modify the code to run `rf"(?<!\w){re.escape(old_name)}(?!\w)"` instead, because there are cases where the shape_expr could be (s1 + s2)*2, and we need to match s1 + s2 to custom name. LMK if this change makes sense.

File: torch/onnx/_internal/exporter/_ir_passes.py (hunk 13 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_ir_passes.py:57]: Done. I need to add a sorting step to mapping because there could be the multiple matching cases, and we want the longest string match, for example, 

{"s1": "seq_len", "s2": "past_seq_len", "s1 + s2": "masked_seq_len"}

This is fine when we check the key and match the value, but it fails when we use it on re.sub. LMK if this change makes sense.

File: torch/onnx/_internal/exporter/_ir_passes.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_ir_passes.py:48]: Maybe we can throw a bunch of example strings to the regex to make sure?

File: torch/onnx/_internal/exporter/_ir_passes.py (hunk 13 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_ir_passes.py:57]: Sorting works I guess!

File: torch/onnx/_internal/exporter/_ir_passes.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_ir_passes.py:48]: I did
